### PR TITLE
[pigeon] Adds non-null object fields

### PR DESCRIPTION
**Pull request by ailtonvivaz**
This PR supports non-nullable fields of classes and enums.

Issue https://github.com/flutter/flutter/issues/59118

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/packages/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by gaaclarke** [comment]
> I'm confused by the description here; non-nullable field support was added in #549. What exactly is this PR doing?

Looks like #549 was only tested on builtin types.  Custom objects have a slightly different flow through the code and this appears to make that work.

File: packages/pigeon/lib/dart_generator.dart [hunk 1]
```diff
@@ -470,15 +470,23 @@ void generateDart(DartOptions opt, Root root, StringSink sink) {
           'final Map<Object$nullTag, Object$nullTag> pigeonMap = <Object$nullTag, Object$nullTag>{};',
         );
         for (final NamedType field in klass.fields) {
+          final String nullsafe = field.type.isNullable ? '?' : '';
```
Comment (gaaclarke) [inline packages/pigeon/lib/dart_generator.dart:473]: Everywhere else in the code we use `isNullSafe` to declare operators, then switch on wether something is nullable or not.  That's reversed here which is a bit harder to follow.

The name "nullsafe" isn't helpful.

**Review by gaaclarke — CHANGES_REQUESTED** [review]
What about objc and java?

File: packages/pigeon/lib/dart_generator.dart (hunk 1 above)
Comment (ailtonvivaz) [reply packages/pigeon/lib/dart_generator.dart]: I've made some changes about it. Could be that?

**Comment by ailtonvivaz** [comment]
> What about objc and java?

I have the follow ideia:
Instead of pass the `DartOptions` to objc and java generators and change every nullable condition to verify if Dart code is nullsafe, we pass the `isNullSafe` to `_RootBuilder` and assign all variables as nullable in case of Dart code be nullsafe.
How about it?

**Comment by gaaclarke** [comment]
> > What about objc and java?
> 
> I have the follow ideia: Instead of pass the `DartOptions` to objc and java generators and change every nullable condition to verify if Dart code is nullsafe, we pass the `isNullSafe` to `_RootBuilder` and assign all variables as nullable in case of Dart code not be nullsafe. How about it?

I'm not sure how that is going to get us objc and java coverage.  We need to do what you did here to verify that non-null objects work correctly on the objc and java generators.

**Comment by ailtonvivaz** [comment]
> > > What about objc and java?
> > 
> > 
> > I have the follow ideia: Instead of pass the `DartOptions` to objc and java generators and change every nullable condition to verify if Dart code is nullsafe, we pass the `isNullSafe` to `_RootBuilder` and assign all variables as nullable in case of Dart code not be nullsafe. How about it?
> 
> I'm not sure how that is going to get us objc and java coverage. We need to do what you did here to verify that non-null objects work correctly on the objc and java generators.

Objc and java generators already test nullable fields.
With https://github.com/flutter/packages/pull/549, it was added the feature to non nullable fields, but looks like it doesn't cover `--no-dart_null_safety`.
Unless this this argument be deprecated, it would be nice cover the pre nullsafe codes.

I thought: If all variables be nullable when the argument `--no-dart_null_safety` is passed, all code will be worked as expected because for Dart without nullsafe, String is same as String? in Dart with nullsafe

**Comment by stuartmorgan** [comment]
What's the status of this PR? Is it waiting for more review, or updates based on the discussion above?

**Comment by ailtonvivaz** [comment]
> What's the status of this PR? Is it waiting for more review, or updates based on the discussion above?

I don't know about the need of java coverage then it would be a revision if the implementation is enough

**Comment by ailtonvivaz** [comment]
Looks like the PR https://github.com/flutter/packages/pull/1543 implements the same thing.
If was the case, this PR could be closed.

**Comment by gaaclarke** [comment]
I see.  I didn't realize that the problem only existed with the Dart generator.  I'll work with stuart to get his PR merged.  What neither PR had was platform tests which we should have.

**Pull request closed without merging**